components/scheduler/scheduler.go

```diff
@@ -0,0 +1,494 @@
+	s.logger.Debug("processBuffer", tag.NewInt("buffer", len(s.Args.State.BufferedStarts)), tag.NewInt("running", len(s.Args.Info.RunningWorkflows)))
+
+	// Make sure we have something to start. If not, we can clear the buffer.
+	req := s.Args.Schedule.Action.GetStartWorkflow()
+	if req == nil || len(s.Args.State.BufferedStarts) == 0 {
+		s.Args.State.BufferedStarts = nil
+		return false
+	}
+
+	isRunning := len(s.Args.Info.RunningWorkflows) > 0
+	tryAgain := false
+
+	action := scheduler.ProcessBuffer(s.Args.State.BufferedStarts, isRunning, s.resolveOverlapPolicy)
+
+	s.Args.State.BufferedStarts = action.NewBuffer
+	s.Args.Info.OverlapSkipped += action.OverlapSkipped
+
+	// Try starting whatever we're supposed to start now
+	allStarts := action.OverlappingStarts
+	if action.NonOverlappingStart != nil {
+		allStarts = append(allStarts, action.NonOverlappingStart)
+	}
+	for _, start := range allStarts {
+		result, err := s.startWorkflow(start, req)
+		metricsWithTag := s.metrics.WithTags(metrics.StringTag(metrics.ScheduleActionTypeTag, metrics.ScheduleActionStartWorkflow))
+		if err != nil {
+			s.logger.Error("Failed to start workflow", tag.NewErrorTag(err))
+			metricsWithTag.Counter(metrics.ScheduleActionErrors.Name()).Record(1)
+			// TODO: we could put this back in the buffer and retry (after a delay) up until
+			// the catchup window. of course, it's unlikely that this workflow would be making
+			// progress while we're unable to start a new one, so maybe it's not that valuable.
+			tryAgain = true
+			continue
+		}
+		metricsWithTag.Counter(metrics.ScheduleActionSuccess.Name()).Record(1)
+		nonOverlapping := start == action.NonOverlappingStart
+		s.recordAction(result, nonOverlapping)
+	}
+
+	// Terminate or cancel if required (terminate overrides cancel if both are present)
+	if action.NeedTerminate {
+		for _, ex := range s.Args.Info.RunningWorkflows {
+			s.terminateWorkflow(ex)
+		}
+	} else if action.NeedCancel {
+		for _, ex := range s.Args.Info.RunningWorkflows {
+			s.cancelWorkflow(ex)
+		}
+	}
+
+	return tryAgain
+}
+
+func (s *Scheduler) startWorkflow(
+	start *schedspb.BufferedStart,
+	newWorkflow *workflowpb.NewWorkflowExecutionInfo,
+) (*schedpb.ScheduleActionResult, error) {
+	nominalTimeSec := start.NominalTime.AsTime().UTC().Truncate(time.Second)
+	workflowID := newWorkflow.WorkflowId
+	if start.OverlapPolicy == enumspb.SCHEDULE_OVERLAP_POLICY_ALLOW_ALL || s.tweakables.AlwaysAppendTimestamp {
+		// must match AppendedTimestampForValidation
+		workflowID += "-" + nominalTimeSec.Format(time.RFC3339)
+	}
+
+	lastCompletionResult, continuedFailure := s.Args.State.LastCompletionResult, s.Args.State.ContinuedFailure
+	if start.OverlapPolicy == enumspb.SCHEDULE_OVERLAP_POLICY_ALLOW_ALL {
+		// ALLOW_ALL runs don't participate in lastCompletionResult/continuedFailure at all
+		lastCompletionResult = nil
+		continuedFailure = nil
+	}
+
+	// TODO(Tianyu): Add callback here
+	req := &schedspb.StartWorkflowRequest{
```

Comment:
you don't need to wrap this? that was just for the local activity

Reply:
Oops